### PR TITLE
Wait for primary pawn blueprint action

### PR DESCRIPTION
**Pull request by ThibNach**
create class to get the primary pawn reference and if it not exist, wait for it

### CODE REVIEW

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+#include <Engine/GameInstance.h>
+
+UCoreExtWaitForPrimaryPawn * UCoreExtWaitForPrimaryPawn::WaitForPrimaryPawn( UObject * world_context_object )
+{
+    ThisClass * action = nullptr;
+
+    if ( auto * world = GEngine->GetWorldFromContextObject( world_context_object, EGetWorldErrorMode::LogAndReturnNull ) )
+    {
+        action = NewObject< ThisClass >();
```
Comment (Frickerson) [inline Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp]: Instead declare and define the var here with auto *
Return at the end of the scope
Return nullptr outside the scope

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#include "CoreExtWaitForPrimaryPawn.h"
+
+#include "GameFramework/GameModeBase.h"
```
Comment (Frickerson) [inline Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp]: <>

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    return action;
+}
+
+void UCoreExtWaitForPrimaryPawn::Activate()
+{
+    Super::Activate();
+
+    if ( const auto * game_instance = WorldPtr.Get()->GetGameInstance() )
+    {
+        if ( auto * player_controller = game_instance->GetPrimaryPlayerController() )
```
Comment (Frickerson) [inline Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp:25]: Cache player controller in member variable
You can then reuse it in the SetReadyToDestroy

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    void SetReadyToDestroy() override;
+
+protected:
+    UPROPERTY( BlueprintAssignable )
+    FOnPrimaryControllerReadyDelegate OnPrimaryPawnDelegate;
+
+private:
+    void Step1_ListenToPrimaryControllerExist( AGameModeBase * game_mode_base, APlayerController * player_controller );
+    void Step2_TryToGetPawn( APlayerController * player_controller );
+    UFUNCTION()
```
Comment (Frickerson) [inline Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h:30]: Whiteline

**Review by Frickerson — CHANGES_REQUESTED** [review]

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp (hunk 1 above)
Comment (ThibNach) [reply Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp]: Done

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp (hunk 2 above)
Comment (ThibNach) [reply Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp]: Done

File: Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp (hunk 3 above)
Comment (ThibNach) [reply Source/CoreExtensions/Private/CoreExtWaitForPrimaryPawn.cpp:25]: Done

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h (hunk 4 above)
Comment (ThibNach) [reply Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h:30]: Done

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <CoreMinimal.h>
+#include <Kismet/BlueprintAsyncActionBase.h>
+
+#include "CoreExtWaitForPrimaryPawn.generated.h"
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE_OneParam( FOnPrimaryControllerReadyDelegate, const APawn *, primary_pawn );
+
+UCLASS()
+class COREEXTENSIONS_API UCoreExtWaitForPrimaryPawn : public UBlueprintAsyncActionBase
```
Comment (TheEmidee) [inline Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: final

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <CoreMinimal.h>
+#include <Kismet/BlueprintAsyncActionBase.h>
+
+#include "CoreExtWaitForPrimaryPawn.generated.h"
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE_OneParam( FOnPrimaryControllerReadyDelegate, const APawn *, primary_pawn );
```
Comment (TheEmidee) [inline Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: FOnPrimaryPawnReadyDelegate

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    void Step1_ListenToPrimaryControllerExist( AGameModeBase * game_mode_base, APlayerController * player_controller );
+    void Step2_TryToGetPawn( APlayerController * player_controller );
+
+    UFUNCTION()
+    void Step3_ListenToPrimaryPawnChanged( APawn * old_pawn, APawn * new_pawn );
+    void Step4_BroadCastPrimaryPawn( const APawn * player_pawn );
+
+    TObjectPtr< APlayerController > PlayerController;
+    TWeakObjectPtr< UWorld > WorldPtr;
+
```
Comment (TheEmidee) [inline Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: remove empty line

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h (hunk 5 above)
Comment (ThibNach) [reply Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: Done

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h (hunk 6 above)
Comment (ThibNach) [reply Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: Done

File: Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h (hunk 7 above)
Comment (ThibNach) [reply Source/CoreExtensions/Public/CoreExtWaitForPrimaryPawn.h]: Done

**Review by Frickerson — APPROVED** [review]

**Pull request merged**